Add multiplayer option to disable score multiplier

timiimit opened the pull request:
This feature spans across osu, osu-server-spectator and osu-web projects. This should be merged only when all PRs are approved.

- [ ] Depends on (ppy/osu-web/pull/10181)

# Changes

## Preview of the change
![image](https://github.com/ppy/osu/assets/32331609/aa237c38-3162-4b8a-ae32-36510ba4e4fd)

## How it works
When this is enabled all players and spectators see the score of everyone without score multiplier applied. When players finish playing everyone also sees the score in results without multiplier applied. Scores get submitted and imported as if they had the multiplier applied. I didn't specifically add checks for newly added `ScoreInfo.IsScoreMultiplierDisabled` on all places, but only where required to make this work. Maybe it would be a good idea to ensure that submitted and imported scores get calculated properly too?


## Tested configurations
### Multiplayer
- 2 players (NoMod, Relax)
- 2 players (Relax, Relax)
- 3 players (NoMod, Relax, Relax)
- 2 players (NoMod, Relax) + 1 spectator
- 1 player (Relax) + 1 spectator
### Solo
- 1 player (NoMod)
- 1 player (Relax)

timiimit committed:
Add room option to disable score multiplier

timiimit committed:
Add ScoreProcessor.TotalScoreWithoutScoreMultiplier

timiimit committed:
Add PlayerConfiguration.ShowScoreWithoutMultiplier

timiimit committed:
Make GameplayScoreCounter show desired score

timiimit committed:
Handle updating correct score in SpectatorScoreProcessor

timiimit committed:
Make multi results screen display scores correctly

timiimit committed:
Handle spectating

timiimit committed:
Remove stupidity

timiimit committed:
Undo test change

timiimit committed:
Remove the need for IsScoreMultiplierApplied

timiimit committed:
Rename ScoreInfo.IsScoreDisplayedWithoutScoreMultiplier

timiimit committed:
Fix code style

timiimit committed:
Merge branch 'master' into multi-score-multiplier-option

timiimit committed:
Handle null check

bdach commented on osu.Game/Rulesets/Scoring/ScoreProcessor.cs:
```diff
@@ -152,7 +152,10 @@ public partial class ScoreProcessor : JudgementProcessor
         private HitObject? lastHitObject;
 
-        private double scoreMultiplier = 1;
+        /// <summary>
+        /// Should generally not be used to change multiplier. Use settter only to override default way of computing multiplier.
+        /// </summary>
+        public double ScoreMultiplier { get; set; } = 1;
```
I'm not sure how to feel about this property being the sole hinge point of this feature but my initial thoughts are "this is not good". If you have to preface a property with "should generally not be used" it should be a big warning sign that maybe things are not entirely structured right. Anything could wipe this change and suddenly all sorts of weird stuff can happen.

If I were to provide an alternative, extensible way of making this work: have a

```csharp
public Func<IEnumerable<Mod>, double> ScoreMultiplierCalculator;
```

field. Initialise it to the default implementation of accumulating `ScoreMultiplier`. Override it to

```csharp
scoreProcessor.ScoreMultiplierCalculator = _ => 1;
```

in the pathways you need that in, and there. No more fear of something funny happening with that. (Unless something else starts messing with the calculator, but at that point, we have bigger problems.)

I guess you could also do a boolean `ApplyMultipliers` flag if needed instead.

timiimit replied:
Good idea

bdach commented on osu.Game/Screens/OnlinePlay/OnlinePlayComposite.cs:
```diff
@@ -89,6 +89,9 @@ public partial class OnlinePlayComposite : CompositeDrawable
         [Resolved(typeof(Room))]
         protected Bindable<bool> AutoSkip { get; private set; }
 
+        [Resolved(typeof(Room))]
+        protected Bindable<bool> NoScoreMultiplier { get; private set; }
```
The naming is all over the place here; I've seen `IsScoreMultiplierDisabled` in `ScoreInfo`, `DisableScoreMultiplier` in `PlayerConfiguration`, and `NoScoreMultiplier` here. Can we pick one variant and stick with it?

bdach commented on osu.Game/Screens/OnlinePlay/Playlists/PlaylistsResultsScreen.cs:
```diff
@@ -182,7 +186,23 @@ private APIRequest createIndexRequest(Action<IEnumerable<ScoreInfo>> scoresCallb
-            var scoreInfos = scoreManager.OrderByTotalScore(scores.Select(s => s.CreateScoreInfo(scoreManager, rulesets, playlistItem, Beatmap.Value.BeatmapInfo))).ToArray();
+            var scoreInfos = scores.Select(s => s.CreateScoreInfo(scoreManager, rulesets, playlistItem, Beatmap.Value.BeatmapInfo)).ToArray();
+
+            if (multiplayerClient?.Room?.Settings.NoScoreMultiplier == true)
+            {
+                // recalculate score without score multiplier
+                var ruleset = Ruleset.Value.CreateInstance();
+                var scoreProcessor = ruleset.CreateScoreProcessor();
+                scoreProcessor.ScoreMultiplier = 1;
+
+                foreach (var scoreInfo in scoreInfos)
+                {
+                    scoreInfo.IsScoreMultiplierDisabled = true;
+                    scoreInfo.TotalScore = scoreProcessor.ComputeScore(Rulesets.Scoring.ScoringMode.Standardised, scoreInfo);
+                }
+            }
+
+            scoreInfos = scoreManager.OrderByTotalScore(scoreInfos).ToArray();
```
Why does this recalculate each score manually, only to then call `OrderByTotalScore()` _again_?

timiimit replied:
I remember there was a reason for it, but it is no longer needed with the changes i did.

bdach reviewed (commented):
Mostly a visual pass with no testing yet

timiimit committed:
Stick to `NoScoreMultiplier` name

timiimit committed:
Add ScoreMultiplierCalculator

timiimit committed:
Add ScoreInfo.ScoreMultiplierCalculator

smoogipoo commented:
Is there an issue or discussion that this PR should be linked to? When would this option be used?

timiimit committed:
Fix bug

timiimit committed:
Merge pull request #1 from timiimit/multi-score-multiplier-option-2

Improve previous implementation of disabling score multiplier

timiimit committed:
Rename readonly field to uppercase

timiimit commented:
As you suggested i added `ScoreMultiplierCalculator` and then expanded on the idea. I made it a generic method that can compute multiplier from `ScoreInfo`. I also added the calculator to `ScoreInfo` so that score itself can determine how multiplier is computed.

I added a static field with default way of computing it to `ScoreInfo`. Perhaps calculator should have a class of its own?

> Mind you, I'm not sure if we want this yet either,...

I'll stop working on it, if it is not desirable. but i have been playing multi on lazer for whole year almost each day and this feature would be really useful constantly. There are people that play multi with conversion mods, HD only, HR only, fun mods, very diverse playstyles and no one really wants to compete with mods. As an example there is someone that plays AR10.6 with DA mod almost 100% of the time and sometimes just for the sake of getting comparable scores everyone else also picks DA with no changes applied.

timiimit committed:
Fix code style

timiimit committed:
Fix formatting

timiimit commented:
I somewhat agree it should be more clear. `Display scores without mod multipliers` is kind of long imo.
Here is how it would look:

![image](https://github.com/ppy/osu/assets/32331609/183c7311-2d17-4c00-aae9-ccf554804033)


It would probably also be good to add a tag (idk yet how those are named in the code) on the room to avoid confusion and make it  visible to everyone. Picture shows where i mean.

![image](https://github.com/ppy/osu/assets/32331609/c6151728-b0f4-47c9-8cab-98d29b3fb784)

But the text you suggested is definitely too long to display here.

monochrome22 commented:
On your screenshot, it says "Mod score multiplier". It should be fine to only use "Mod multiplier"; everyone should know what it refers to and that's also the term used in the [wiki](https://osu.ppy.sh/wiki/en/Gameplay/Game_modifier/Mod_multiplier). Also, it should be plural, "multipliers". I don't think my sentence is too long, it's only 2 more characters than the intro skip text. More opinions on this are welcome.
For the badge, probably just "No mod multipliers". It would be good to also have a tooltip with additional explanation, but you should wait for more opinions before adding it, because other badges don't have tooltips.

peppy commented:
> Mind you, I'm not sure if we want this yet either, but there is the cause of this entire series.

I personally would like to see this one. It's along the same lines of allowing players to play different difficulties and make multiplayer more friends to groups of differing skill levels. It's pretty common in arcade rhythm games and makes playing with friends or randoms so much more enjoyable.

timiimit committed:
Change checkbox text

timiimit committed:
Merge branch 'ppy:master' into multi-score-multiplier-option

timiimit committed:
Add `NoScoreMultiplierPill`

timiimit commented:
I changed text as suggested and added a badge to the room which is shown when this option is enabled and hidden when disabled.

![image](https://github.com/ppy/osu/assets/32331609/cd85bce5-b10b-45d4-bd6e-96a050e7dd7c)

timiimit committed:
Remove unused namespace

peppy commented:
> In my opinion, "Disable mod score multiplier" should be rephrased to "Display scores without mod multipliers". As it is now, players might get the impression that they can submit scores without multipliers by using this option in multiplayer. It should be made more clear that this option is only a visual change.

This makes no sense. In the context of a multiplayer game, it's *not* a visual-only change. It changes the outcome of the room / tournament / whatever.

I would write this as "No mod multipliers" or similar.

monochrome22 commented:
The option doesn't actually disable mod multipliers, it just shows what the scores would look like if they didn't have multipliers applied, they are still being submitted with multipliers applied, that's why I said it's a visual change. It's true that in the context of the lobby, it's not a visual change, but in the context of the entire game, it is a visual change.
If you feel that players misinterpreting the functionality of the option is not a concern, then feel free to use "No mod multipliers".

timiimit commented:
I agree that "No mod multipliers" would be better, since this option is in the context of multi.
I suppose initially some may think that scores would submit as displayed, but its probably just going to be one of the things that people will learn eventually.

timiimit committed:
Merge branch 'ppy:master' into multi-score-multiplier-option

timiimit committed:
Change checkbox text

peppy commented:
> I suppose initially some may think that scores would submit as displayed, but its probably just going to be one of the things that people will learn eventually.

I mean, right now multiplayer scores don't "submit" for solo play. They are isolated to the multiplayer system.

peppy commented:
As a heads-up, we should probably hold off merging this until the score v2 changes are applied (#23638), as I foresee conflicts and would rather that PR has precedence.

timiimit committed:
Merge branch 'master' into multi-score-multiplier-option

timiimit committed:
Use newly merged `OnlinePlayPill`

timiimit committed:
Remove empty line

timiimit committed:
Merge branch 'master' into multi-score-multiplier-option

timiimit commented:
Since ScoreV2 is not recalculatable, I still need to make it calculate score with and without multiplier.

bdach commented on osu.Game/Online/Spectator/SpectatorScoreProcessor.cs:
```diff
@@ -49,7 +50,7 @@ public partial class SpectatorScoreProcessor : Component
         public IReadOnlyList<Mod> Mods => scoreInfo?.Mods ?? Array.Empty<Mod>();
 
-        public Func<ScoringMode, long> GetDisplayScore => mode => scoreInfo?.GetDisplayScore(mode) ?? 0;
+        public Func<ScoringMode, bool, long> GetDisplayScore => (mode, noScoreMultiplier) => scoreInfo?.GetDisplayScore(mode, noScoreMultiplier) ?? 0;
```
I have a bit of a dislike for `Func<>`s that have `bool`s inside with no real clear meaning. I would propose declaring a [named delegate type](https://learn.microsoft.com/en-US/dotnet/csharp/programming-guide/delegates/) for this and properly naming the parameters of said delegate to describe them without needing to look up a random implementation to find the lambda param name.

bdach commented on osu.Game/Scoring/Legacy/ScoreInfoExtensions.cs:
```diff
@@ -12,14 +12,21 @@ namespace osu.Game.Scoring.Legacy
-        private static long getDisplayScore(int rulesetId, long score, ScoringMode mode, IReadOnlyDictionary<HitResult, int> maximumStatistics)
+        private static long getDisplayScore(int rulesetId, long score, double scoreMultiplier, bool noScoreMultiplier, ScoringMode mode, IReadOnlyDictionary<HitResult, int> maximumStatistics)
         {
+            if (noScoreMultiplier)
+            {
+                // This might result in a minor inaccuracy of score (+-1 point) because the passed in score is already rounded.
+                // This shouldn't be much of a problem because noScoreMultiplier is only a visual effect meant for fun.
+                score = (long)Math.Round(score / scoreMultiplier);
+            }
```
Having to pass `scoreMultiplier` and `noScoreMultiplier` and having to _undo_ an already-applied multiplier feels completely backwards. Much less so that this comment states that the undo _cannot be done accurately_ due to rounding applied.

timiimit replied:
I don't think having a minor inaccuracy is problematic. This feature implements only a visual difference in scores. Having this feature coded alongside `ScoringMode` as final step in calculating score that is displayed makes sense to me. Classic scoring is also just a fake front for users anyway, since internally calculations are done the same way as standardised scoring.

I made `ScoreProcessor` calculate proper score, but `ScoreInfo` still requires score multiplier to be unapplied.

bdach replied:
> Do you think it be better to ditch `GetDisplayScore`?

I don't know how to answer this question because I don't know why the implementation has swung in this entire different direction. The initial version wherein the application of the multipliers (or lack thereof) was coordinated by `ScoreProcessor` and `MultiplayerPlayer` was miles better because it did not weirdly couple components like what this new variant is doing. Is there a constraint that makes that old implementation unviable that I am missing here?

timiimit replied:
I forgot what the reason was, so I tried to work on previous implementation again [here](https://github.com/ppy/osu/compare/master...timiimit:osu:multi-score-multiplier-option-rework?expand=1). Actually, I probably didn't think enough about what continuing with old implementation would require. ScoreV2 update made all code unfamiliar and it seemed too hard to continue so I switched up my approach. In reality I only needed to add a couple of lines...

The actual reason for new implementation was that I couldn't figure out a way to recalculate score with `ScoreProcessor` anymore. In linked branch I still can't recalculate when score is retrieved from server which is done on the results screen. Before ScoreV2 update, `ScoreProcessor.CalculateScore()` could return score but this method is now gone. This makes implementation still suffer from unappyling score multiplier in `PlaylistResultsScreen`.

Let me know if I should continue with linked branch.

bdach commented on osu.Game/Scoring/ScoreManager.cs:
```diff
@@ -41,10 +44,11 @@ public override bool PauseImports
 
         public ScoreManager(RulesetStore rulesets, Func<BeatmapManager> beatmaps, Storage storage, RealmAccess realm, IAPIProvider api,
-                            OsuConfigManager configManager = null)
+                            OsuConfigManager configManager = null, MultiplayerClient multiplayerClient = null)
```
I vehemently disagree with roping `MultiplayerClient` into `ScoreManager`. `ScoreManager` should not know anything about multiplayer, period.

timiimit replied:
Shouldn't `ScoreManager` have everything that could affect the score?

peppy replied:
Extract the piece out that it needs to know about, such as a single bool, at most. You can't just pass full classes – it creates  weird cross-component dependencies that we try to avoid.

bdach commented on osu.Game/Scoring/ScoreManager.cs:
```diff
@@ -88,7 +92,7 @@ public IEnumerable<ScoreInfo> OrderByTotalScore(IEnumerable<ScoreInfo> scores)
         /// <param name="score">The <see cref="ScoreInfo"/> to retrieve the bindable for.</param>
         /// <returns>The bindable containing the total score.</returns>
-        public Bindable<long> GetBindableTotalScore([NotNull] ScoreInfo score) => new TotalScoreBindable(score, configManager);
+        public Bindable<long> GetBindableTotalScore([NotNull] ScoreInfo score) => new TotalScoreBindable(score, configManager, multiplayerClient.APIRoom);
```
Same goes for this. `GetBindableTotalScore()` should not need to know anything about any rooms, period.

bdach reviewed (changes requested):
Aside from the numerous test failures and failing code quality inspections, this implementation feels much worse than previously

timiimit commented:
Score retrieved from `ScoreProcessor` is now handled properly without undoing multiplier, but `ScoreInfo` only has final `long TotalScore` with `ScoreMultiplier` applied so exact value from before `ScoreMultiplier` was applied is impossible to get because of rounding.

`ScoreInfo.GetDisplayScore` is used to see others' live scores and on results screen. I will try to make it so that `TotalScore` in `FrameDataBundle` contains score without score multiplier.

bdach commented on osu.Game/Online/Multiplayer/MultiplayerClient.cs:
```diff
@@ -143,7 +143,29 @@ public virtual bool IsHost
+            }
+        }
+
+        public Bindable<bool> NoScoreMultiplier = new Bindable<bool>(false);
```
This still baffles me. The bindable is essentially global state for a given client instance. Why does it even need to exist in this way? Why does a communication component concern itself with scoring? The _consumers_ of the communication component - namely `MultiplayerPlayer` - should be the one caring about scoring.

bdach commented on osu.Game/Scoring/ScoreInfo.cs:
```diff
@@ -184,6 +184,12 @@ public ScoreInfo DeepClone()
         [Ignored]
         public bool IsLegacyScore => Mods.OfType<ModClassic>().Any();
 
+        /// <summary>
+        /// Score multiplier of this score. Value will be correct only if <see cref="mods"/> is non-null.
+        /// </summary>
+        [Ignored]
+        public double ScoreMultiplier { get; private set; } = 1;
```
If you have to preempt that a property is only valid to read in specific circumstances, either enforce that the circumstances are met via an exception or something, or, preferably, just don't add it at all, and inline it in the specific circumstance you need to use it in.

timiimit committed:
Merge branch 'master' into multi-score-multiplier-option-rework

timiimit committed:
Apply score multiplier before import

timiimit committed:
Unapply score multiplier in results screen

bdach commented on osu.Game/Rulesets/Scoring/ScoreProcessor.cs:
```diff
@@ -40,6 +40,11 @@ public partial class ScoreProcessor : JudgementProcessor
+        /// <summary>
+        /// The current score without <see cref="ScoreMultiplier"/> applied.
+        /// </summary>
+        public readonly BindableDouble BaseScore = new BindableDouble { MinValue = 0 };
```
I'm still really lost as to why all this is exposing `BaseScore` and `ScoreMultiplier` when we already _had_ an implementation that was completely constrained to `ScoreProcessor` in https://github.com/ppy/osu/pull/23510#discussion_r1193185796.

I want this score multiplier disable stuff to touch as little places as possible. It should be in `ScoreProcessor`, and it should be in the realm database so that replays can be played back correctly (arguable depending on expectations and requirements that are still not clear to me on that front). Nothing else should need to concern itself about any of this.

timiimit replied:
I reset implementation back to the way it was.

Hmm, I last time I checked locally saved score doesn't contin total score in replay frames. Only final score is handled currently.

timiimit committed:
Merge branch 'master' into multi-score-multiplier-option

timiimit commented:
I reset branch to commit `de6a51d0d9f4b225c89af1b080c6c878bca794b1` and then merged with branch mentioned [here](https://github.com/ppy/osu/pull/23510#discussion_r1224662252).

timiimit committed:
Fix code quality

bdach commented:
I looked at this diff again and saw things being put in places where I don't think they should be (the `ScoreMultiplierCalculator` thing in `ScoreInfo` - why is it there at all? - and in `PlayerConfiguration` - does it need to be there? can't it be added manually in the multiplayer/spectator players themselves? does every single player need to be aware of this flag?)

I don't know. I give up. Need someone else to look at this diff with a fresh outlook because I don't know what's happening here at this stage and I don't know whether I'm seeing actual problems at this stage.

ReinIsNOTaDev commented:
is this pr dead?

bdach commented:
Probably. It needs to have conflicts resolved, and considerations made for things that we did to get lazer leaderboards live - at the least. And I didn't like the code quality here before either, so this is likely in need of a full rewrite if we want it to go forward.

timiimit commented:
I kind of stopped working on this, as I no longer liked the direction I originally took.

I thought it would be better if each individual was able to choose whether to see scores with or without multiplier instead of having room settings checkbox. This would make each player happy because they would see scores in their preferred  way. But the downside is that people would be seeing scores in two different ways, which would surely be confusing. Especially where this option would change the order of players in final results screen, so there could be 2 distinct winners based on which option you use. So something would have to make it visually apparent as to why scores appear different among players.

I suppose this already is a similar problem as with classic and standard scoring. But there you can usually tell who has classic enabled because scores are so much higher (and I assume order of scores is preserved).

I also think I saw in PRs or heard from somewhere that score & score multiplier are now separated internally. And that makes it easier to now implement this.

bdach commented:
> I thought it would be better if each individual was able to choose whether to see scores with or without multiplier instead of having room settings checkbox

Definitely not happening.

> I also think I saw in PRs or heard from somewhere that score & score multiplier are now separated internally

We're not there yet but this sort of thing is happening to facilitate future mod multiplier rebalances.
